keps/sig-storage/2639-secret-configmap-protection/README.md

```diff
@@ -0,0 +1,652 @@
+Users need to add below to enable this feature per resource:
+- `kubernetes.io/enable-secret-protection` annotation to Secret
+- `kubernetes.io/enable-configmap-protection` annotation to ConfigMap
+
+Basic flows of Secret protection controller and ConfigMap protection controller are as follows:
+
+- Secret protection controller:
+  - The controller watches `Pod`, `PersistentVolume`, `VolumeSnapshotContent`, `Deployment`, `StatefulSet`, `DaemonSet`, and `CronJob`
+    - Create/Update event:
+      - For all the `Secret`s referenced by the resource:
+        1. Update the dependency graph (If the resource is using a Secret, mark the Secret that the resource is using it)
+        2. If there are any updates to the Secret in the dependency graph, add the key for the Secret to add-lien-queue
+    - Delete event:
+      - For all the `Secret`s referenced by the resource:
+        1. Update the dependency graph (If the resource is using a Secret, unmark the Secret that the resource is using it)
+        2. If there are no resource using the Secret in the dependency graph, add the key for the Secret to delete-lien-queue
+  - The controller periodically checks `Pod`, `PersistentVolume`, `VolumeSnapshotContent`, `Deployment`, `StatefulSet`, `DaemonSet`, and `CronJob`
+    - For each resource:
+      1. Update the dependency graph (If the resource is using a Secret, mark the Secret that the resource is using it)
+      2. If there are any updates to the Secret in the dependency graph, add the key for the Secret to add-lien-queue
+  - The controller periodically checks `Secret`
+    - For each `Secret`:
+      - Check if the `Secret` have `kubernetes.io/enable-secret-protection: "yes"` annotation:
+        - If yes:
+          1. Check API server if the resources marked as using the secret in the dependency graph are still using the secret, and update the graph
+          2. Check if there is no using resources in the graph any more
+            - If yes: Add the key for the secret to delete-lien-queue
+            - Otherwise: Add the key for the secret to add-lien-queue
+        - Otherwise: Add the key for the secret to delete-lien-queue
+  - The controller gets a key for a `Secret` from add-lien-queue:
+    1. If the `Secret` doesn't have `kubernetes.io/enable-secret-protection: "yes"` annotation, do nothing
+    2. Check API server if the `Secret` is actually used by one of the resources marked in the dependency graph, and if used, add `kubernetes.io/secret-protection` lien to the `Secret`
+  - The controller gets a key for a `Secret` from delete-lien-queue:
+    1. If the `Secret` doesn't have `kubernetes.io/enable-secret-protection: "yes"` annotation, delete `kubernetes.io/secret-protection` lien from the `Secret`
+    2. Check API server if the `Secret` is actually not used by any of the resources, and if not used, delete `kubernetes.io/secret-protection` lien from the `Secret`
+
+- ConfigMap protection controller:
```

Comment:
Can we just say that ConfigMap controller does pretty much exactly the same as Secret controller?
This whole text doesn't bring any additional value.